pkg/signup/service/signup_service.go

```diff
@@ -131,6 +146,53 @@ func (s *ServiceImpl) newUserSignup(ctx *gin.Context) (*toolchainv1alpha1.UserSi
 	return userSignup, nil
 }
 
+/*
+IsPhoneVerificationRequired determines whether phone verification is required
+
+Returns true in the following cases:
+1. Captcha configuration is disabled
+2. The captcha token is invalid
+3. Captcha failed with an error or the assessment failed
+
+Returns false in the following cases:
+1. Overall verification configuration is disabled
+2. Captcha is enabled and the assessment is successful
+*/
+func IsPhoneVerificationRequired(captchaChecker captcha.Assessor, ctx *gin.Context) (bool, float32) {
+	cfg := configuration.GetRegistrationServiceConfig()
+	if !cfg.Verification().Enabled() {
+		return false, -1
+	}
+
+	if !cfg.Verification().CaptchaEnabled() {
+		return true, -1
+	}
+
+	if ctx == nil || ctx.Request == nil {
+		log.Error(ctx, nil, "no request in context")
+		return true, -1
+	}
+
+	captchaToken, exists := ctx.Request.Header["Recaptcha-Token"]
+	if !exists || len(captchaToken) != 1 {
+		log.Error(ctx, nil, "no valid captcha token found in request header")
+		return true, -1
+	}
+
+	score, err := captchaChecker.CompleteAssessment(ctx, cfg, captchaToken[0])
+	if err != nil {
+		log.Error(ctx, err, "signup assessment failed")
+		return true, -1
+	}
+
+	if score < cfg.Verification().CaptchaScoreThreshold() {
+		log.Error(ctx, fmt.Errorf("the risk analysis score did not meet the expected threshold"), "signup assessment failed")
```

Comment:
It's not an error. Should be `log.Info()` instead if we want to log it. Plus the actual score and the threshold could be useful in the log message.

Reply:
Good point. Done in 900c1ee7cff3b6cf8288262469e6c3ce4354230c